Abins2D: validate chopper parameters

ajjackson opened the pull request:
### Description of work

Improve Abins2D validator; for chopper instruments, attempt to get a Pychop resolution value and ensure there is neutron transmission. If not, better to raise a validation error now than  _after_ calculating intensities.

Closes #38085

**Report to:** Hamish Cavaye

### To test:

A unit test is included.

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?
- If the PR author isn’t in the `mantid-developers` or `mantid-contributors` teams, add a review comment `rerun ci` to authorize/rerun the CI

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

ajjackson committed:
Abins2D: validate chopper parameters

ajjackson committed:
Add release note

ajjackson committed:
Improve docstring, type annotations for new validation function

adriazalvarez reviewed (changes requested):
I have only found one issue,  other than that it looks very good and I'd be happy to approve once is resolved.

ajjackson commented on scripts/abins/instruments/pychop.py:
```diff
@@ -68,3 +69,42 @@ def _polyfit_resolution(self, n_values=40, order=4):
+
+    EPSILON = 1e-8  # Calculation range is slightly reduced to avoid considering stationary neutrons
+
+    energy = float(e_i)
```
I think `e_i` is already checked: `abins.abinsalgorithm.validate_e_init()` is called, which includes

```python
    if not isinstance(float(e_init), numbers.Real):
        issues["IncidentEnergy"] = "Incident energy must be a real number"
        return issues

```

For `ChopperFrequency` an empty value is supposed to select an instrument-specific default and there is already a validator checking that a default is available if you leave it empty. But I forgot to support this scenario in `validate_pychop_params`, will add that now! Perhaps I can move more of the logic into that function to improve testing.

ajjackson committed:
Move more Abins2D/pychop validation to pure function

In response to review by @adriazalvarez

- This lets us test more cleanly
- and add another check while we're at it (for values that don't cast
  to int properly)
- Also fix a bug (empty-string handling was broken by previous changes)

adriazalvarez commented on scripts/abins/instruments/pychop.py:
```diff
@@ -91,6 +92,23 @@ def validate_pychop_params(
+        return {
+            "ChopperFrequency": f"This chopper frequency is not valid for the instrument {name}. "
+            "Valid frequencies: " + ", ".join([str(freq) for freq in allowed_frequencies])
+        }
```
This block shouldn't execute if ``chopper_frequency==''`` and ``default_frequency`` is not ``None`` . I'm guessing when you don't provide chopper frequency it should take the default if the instrument requires one, is this correct?

ajjackson committed:
Rework pychop frequency validation logic

Some cases were not treated correctly, e.g. default frequency with
incompatible incident energy.

adriazalvarez reviewed (approved):
All good to me. Thanks for the changes 👍
